x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java

```diff
@@ -1320,14 +1358,16 @@ default SourceInputStreamFactory sharedInputStreamFactory(List<SparseFileTracker
          * @param length of data to fetch
          * @param progressUpdater consumer to invoke with the number of copied bytes as they are written in cache.
          *                        This is used to notify waiting readers that data become available in cache.
+         * @param completionListener listener that has to be called when the callback method completes
          */
         void fillCacheRange(
             SharedBytes.IO channel,
             int channelPos,
             @Nullable SourceInputStreamFactory streamFactory,
             int relativePos,
             int length,
-            IntConsumer progressUpdater
+            IntConsumer progressUpdater,
+            ActionListener<Void> completionListener
```

Comment:
I wonder if we could skip the change to the `fillCacheRange` method in this PR and focus on just making `SourceInputStreamFactory#create` async. The PR and the associated one in the other repo would become quite a bit simpler. More importantly, for the immediate task at hand, i.e. fetching data from indexing node asynchronously, the blocking concern is only when creating the `InputStream`. Reading from it subsequently in `fillCacheRange` is from an in-memory `BytesReferenceStreamInput` which does not need to be async. For opening and reading input stream from the object store, things can be as is for the time being until s3 client is upgraded to be asynchronous (or maybe when we need to make azure and gcs work). 

I do think we likely need async `fillCacheRange` at some point. My suggestion is that we can asyncify it separately in a future PR which would at least make coding and reviewing the associated PR in the other repo simpler.

Reply:
Yeah, in the retrospect it would have been easier just to focus on making `SourceInputStreamFactory` async, but I think it's better to proceed with making `fillCacheRange` async, since we already did a lot of work in making it happen.